csrc/xpu_ops.cpp

```diff
@@ -0,0 +1,111 @@
+#include <common.h>
+#include <xpu_kernels.h>
+#include <xpu_ops.h>
+
+template <typename T, int DATA_TYPE>
+void dequantizeBlockwise(float *code, unsigned char *A, float *absmax, T *out,
+                         int blocksize /*block-quant-size*/, const int n,
+                         sycl::queue *stream) {
+  auto &queue = *stream;
+  const int workgroup_size = 128;
+  const int num_per_th = 4;
+  const int tile_size = workgroup_size * num_per_th;
+  if (DATA_TYPE > 0) {
+    const int workgroup_num = (n + tile_size - 1) / tile_size / 2;
+    sycl::range<1> local_range{(size_t)workgroup_size};
+    sycl::range<1> global_range{(size_t)workgroup_num * (size_t)workgroup_size};
+    kDequantizeBlockwise<T, tile_size, num_per_th, DATA_TYPE> kfn(
+        code, A, absmax, out, blocksize / 2, n);
+    sycl_kernel_submit<decltype(kfn), 1, 32>(
+        sycl::nd_range<1>(sycl::range<1>(global_range),
+                          sycl::range<1>(local_range)),
+        queue, kfn);
+  } else {
+    const int workgroup_num = (n + tile_size - 1) / tile_size;
+    sycl::range<1> local_range{(size_t)workgroup_size};
+    sycl::range<1> global_range{(size_t)workgroup_num * (size_t)workgroup_size};
+    kDequantizeBlockwise<T, tile_size, num_per_th, DATA_TYPE> kfn(
+        code, A, absmax, out, blocksize, n);
+    sycl_kernel_submit<decltype(kfn), 1, 32>(
+        sycl::nd_range<1>(sycl::range<1>(global_range),
+                          sycl::range<1>(local_range)),
+        queue, kfn);
+  }
+}
+
+template <typename T, int BITS>
+void gemv_4bit_inference(int m, int n, int k, T *A, unsigned char *B,
+                         float *absmax, float *datatype, T *out, int lda,
+                         int ldb, int ldc, int blocksize, sycl::queue *stream) {
+
+  auto &queue = *stream;
+
+  size_t subgroup_size = 32;
+  size_t workgroup_size = subgroup_size * 4;
+  size_t workgroup_num = (n + 3) / 4;
+
+  const int THREADS = 128;  // workgroup_size;
+  const int SUBG_SIZE = 32; // subgroup_size;
```

Comment:
I suggest not using hard-coded sub group size, but query sub group size on the fly, which is more compatible for different Intel GPU platforms.